cmd/bundle/generate/utils.go

```diff
@@ -0,0 +1,138 @@
+package generate
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/databricks/cli/libs/cmdio"
+	"github.com/databricks/cli/libs/notebook"
+	"github.com/databricks/databricks-sdk-go"
+	"github.com/databricks/databricks-sdk-go/service/jobs"
+	"gopkg.in/yaml.v3"
+)
+
+type notebookDownloader struct {
+	notebooks map[string]io.ReadCloser
+	w         *databricks.WorkspaceClient
+	outputDir string
+}
+
+func saveConfigToFile(ctx context.Context, data any, filename string, force bool) error {
+	// check that file exists
+	info, err := os.Stat(filename)
+	if err == nil {
+		if info.IsDir() {
+			return fmt.Errorf("%s is a directory", filename)
+		}
+		if !force {
+			return fmt.Errorf("%s already exists. Use --force to overwrite", filename)
+		}
+	}
+
+	file, err := os.Create(filename)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	err = encode(data, file)
+	if err != nil {
+		return err
+	}
+
+	cmdio.LogString(ctx, fmt.Sprintf("Job configuration successfully saved to %s", filename))
+	return nil
+}
+
+func encode(data any, w io.Writer) error {
+	enc := yaml.NewEncoder(w)
+	enc.SetIndent(2)
+	return enc.Encode(data)
+}
+
+func (n *notebookDownloader) DownloadInMemory(ctx context.Context, task *jobs.Task) error {
+	if task.NotebookTask == nil {
+		return nil
+	}
+
+	info, err := n.w.Workspace.GetStatusByPath(ctx, task.NotebookTask.NotebookPath)
+	if err != nil {
+		return err
+	}
+
+	ext := notebook.GetExtensionByLanguage(info)
+
+	filename := path.Base(task.NotebookTask.NotebookPath) + ext
+	targetPath := filepath.Join(n.outputDir, filename)
+
+	// if not yet downloaded, download first
+	if _, ok := n.notebooks[targetPath]; !ok {
+		reader, err := n.w.Workspace.Download(ctx, task.NotebookTask.NotebookPath)
+		if err != nil {
+			return err
+		}
+		n.notebooks[targetPath] = reader
```

Comment:
This is tricky; by deferring the read and close we're keeping the underlying connections open. If you have 100 notebooks, these calls may start to fail due to rate limits or maximum # of open connections.

I think we're OK to read their contents into memory given that notebooks have a maximum size in the workspace already, which in turn translates into an upper bound of memory usage here in the client.

Alternatively, we can store the notebook paths to download, and then make these download API calls (potentially with an `errgroup` for parallelism) upon flushing to disk.

Reply:
The `GetStatusByPath` call succeeds, so I believe we're safe to assume that the download calls will work.